File: sdks/php/src/Module/HashManager.php
<?php

declare(strict_types=1);

namespace Synap\SDK\Module;

use Synap\SDK\SynapClient;

/**
 * Hash data structure operations (Redis-compatible)
 */
class HashManager
{
    public function __construct(
        private SynapClient $client
    ) {
    }

    /**
     * Set field in hash
     *
     * @param string $key Hash key
     * @param string $field Field name
     * @param string|int|float $value Field value
     */
    public function set(string $key, string $field, string|int|float $value): bool
    {
        $response = $this->client->execute('hash.set', $key, [
            'field' => $field,
            'value' => (string) $value,
        ]);

        return (bool) ($response['success'] ?? false);
    }

    /**
     * Get field from hash
     */
    public function get(string $key, string $field): ?string
    {
        $response = $this->client->execute('hash.get', $key, ['field' => $field]);

        return $response['value'] ?? null;
    }

    /**
     * Get all fields and values from hash
     *
     * @return array<string, string>
     */
    public function getAll(string $key): array
    {
        $response = $this->client->execute('hash.getall', $key);

        return $response['fields'] ?? [];
    }

    /**
     * Delete field from hash
     */
    public function delete(string $key, string $field): int
    {
        $response = $this->client->execute('hash.del', $key, ['field' => $field]);

        return (int) ($response['deleted'] ?? 0);
    }

    /**
     * Check if field exists in hash
     */
    public function exists(string $key, string $field): bool
    {
        $response = $this->client->execute('hash.exists', $key, ['field' => $field]);

        return (bool) ($response['exists'] ?? false);
    }

    /**
     * Get all field names in hash
     *
     * @return list<string>
     */
    public function keys(string $key): array
    {
        $response = $this->client->execute('hash.keys', $key);

        return $response['fields'] ?? [];
    }

    /**
     * Get all values in hash
     *
     * @return list<string>
     */
    public function values(string $key): array
    {
        $response = $this->client->execute('hash.values', $key);

        return $response['values'] ?? [];
    }

    /**
     * Get number of fields in hash
     */
    public function len(string $key): int
    {
        $response = $this->client->execute('hash.len', $key);

        return (int) ($response['length'] ?? 0);
    }

    /**
     * Set multiple fields in hash
     *
     * @param array<string, string|int|float> $fields
     */
    public function mset(string $key, array $fields): bool
    {
        $strFields = array_map('strval', $fields);
        $response = $this->client->execute('hash.mset', $key, ['fields' => $strFields]);

        return (bool) ($response['success'] ?? false);
    }

    /**
     * Get multiple fields from hash
     *
     * @param list<string> $fields
     * @return array<string, string|null>
     */
    public function mget(string $key, array $fields): array
    {
        $response = $this->client->execute('hash.mget', $key, ['fields' => $fields]);

        return $response['values'] ?? [];
    }

    /**
     * Increment field value by integer
     */
    public function incrBy(string $key, string $field, int $increment): int
    {
        $response = $this->client->execute('hash.incrby', $key, [
            'field' => $field,
            'increment' => $increment,
        ]);

        return (int) ($response['value'] ?? 0);
    }

    /**
     * Increment field value by float
     */
    public function incrByFloat(string $key, string $field, float $increment): float
    {
        $response = $this->client->execute('hash.incrbyfloat', $key, [
            'field' => $field,
            'increment' => $increment,
        ]);

        return (float) ($response['value'] ?? 0.0);
    }

    /**
     * Set field only if it doesn't exist
     */
    public function setNX(string $key, string $field, string|int|float $value): bool
    {
        $response = $this->client->execute('hash.setnx', $key, [
            'field' => $field,
            'value' => (string) $value,
        ]);

        return (bool) ($response['created'] ?? false);
    }
}

File: sdks/csharp/src/Synap.SDK/Modules/HashManager.cs
using System.Text.Json;

namespace Synap.SDK.Modules;

/// <summary>
/// Hash data structure operations (Redis-compatible).
/// </summary>
public sealed class HashManager
{
    private readonly SynapClient _client;

    public HashManager(SynapClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Set field in hash.
    /// </summary>
    public async Task<bool> SetAsync(
        string key,
        string field,
        string value,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>
        {
            ["field"] = field,
            ["value"] = value
        };

        using var response = await _client.ExecuteAsync("hash.set", key, data, cancellationToken).ConfigureAwait(false);
        return response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean();
    }

    /// <summary>
    /// Get field from hash.
    /// </summary>
    public async Task<string?> GetAsync(string key, string field, CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?> { ["field"] = field };
        using var response = await _client.ExecuteAsync("hash.get", key, data, cancellationToken).ConfigureAwait(false);

        if (response.RootElement.TryGetProperty("value", out var value))
        {
            return value.GetString();
        }

        return null;
    }

    /// <summary>
    /// Get all fields and values from hash.
    /// </summary>
    public async Task<Dictionary<string, string>> GetAllAsync(string key, CancellationToken cancellationToken = default)
    {
        using var response = await _client.ExecuteAsync("hash.getall", key, null, cancellationToken).ConfigureAwait(false);

        if (response.RootElement.TryGetProperty("fields", out var fields))
        {
            return JsonSerializer.Deserialize<Dictionary<string, string>>(fields.GetRawText()) ?? new();
        }

        return new Dictionary<string, string>();
    }

    /// <summary>
    /// Delete field from hash.
    /// </summary>
    public async Task<int> DeleteAsync(string key, string field, CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?> { ["field"] = field };
        using var response = await _client.ExecuteAsync("hash.del", key, data, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("deleted", out var deleted) ? deleted.GetInt32() : 0;
    }

    /// <summary>
    /// Check if field exists in hash.
    /// </summary>
    public async Task<bool> ExistsAsync(string key, string field, CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?> { ["field"] = field };
        using var response = await _client.ExecuteAsync("hash.exists", key, data, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("exists", out var exists) && exists.GetBoolean();
    }

    /// <summary>
    /// Get all field names in hash.
    /// </summary>
    public async Task<List<string>> KeysAsync(string key, CancellationToken cancellationToken = default)
    {
        using var response = await _client.ExecuteAsync("hash.keys", key, null, cancellationToken).ConfigureAwait(false);

        if (response.RootElement.TryGetProperty("fields", out var fields))
        {
            return JsonSerializer.Deserialize<List<string>>(fields.GetRawText()) ?? new();
        }

        return new List<string>();
    }

    /// <summary>
    /// Get all values in hash.
    /// </summary>
    public async Task<List<string>> ValuesAsync(string key, CancellationToken cancellationToken = default)
    {
        using var response = await _client.ExecuteAsync("hash.values", key, null, cancellationToken).ConfigureAwait(false);

        if (response.RootElement.TryGetProperty("values", out var values))
        {
            return JsonSerializer.Deserialize<List<string>>(values.GetRawText()) ?? new();
        }

        return new List<string>();
    }

    /// <summary>
    /// Get number of fields in hash.
    /// </summary>
    public async Task<int> LenAsync(string key, CancellationToken cancellationToken = default)
    {
        using var response = await _client.ExecuteAsync("hash.len", key, null, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("length", out var length) ? length.GetInt32() : 0;
    }

    /// <summary>
    /// Set multiple fields in hash.
    /// </summary>
    public async Task<bool> MSetAsync(
        string key,
        Dictionary<string, string> fields,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?> { ["fields"] = fields };
        using var response = await _client.ExecuteAsync("hash.mset", key, data, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean();
    }

    /// <summary>
    /// Get multiple fields from hash.
    /// </summary>
    public async Task<Dictionary<string, string?>> MGetAsync(
        string key,
        IReadOnlyList<string> fields,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?> { ["fields"] = fields };
        using var response = await _client.ExecuteAsync("hash.mget", key, data, cancellationToken).ConfigureAwait(false);

        if (response.RootElement.TryGetProperty("values", out var values))
        {
            return JsonSerializer.Deserialize<Dictionary<string, string?>>(values.GetRawText()) ?? new();
        }

        return new Dictionary<string, string?>();
    }

    /// <summary>
    /// Increment field value by integer.
    /// </summary>
    public async Task<long> IncrByAsync(
        string key,
        string field,
        long increment,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>
        {
            ["field"] = field,
            ["increment"] = increment
        };

        using var response = await _client.ExecuteAsync("hash.incrby", key, data, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("value", out var value) ? value.GetInt64() : 0;
    }

    /// <summary>
    /// Increment field value by float.
    /// </summary>
    public async Task<double> IncrByFloatAsync(
        string key,
        string field,
        double increment,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>
        {
            ["field"] = field,
            ["increment"] = increment
        };

        using var response = await _client.ExecuteAsync("hash.incrbyfloat", key, data, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("value", out var value) ? value.GetDouble() : 0.0;
    }

    /// <summary>
    /// Set field only if it doesn't exist.
    /// </summary>
    public async Task<bool> SetNXAsync(
        string key,
        string field,
        string value,
        CancellationToken cancellationToken = default)
    {
        var data = new Dictionary<string, object?>
        {
            ["field"] = field,
            ["value"] = value
        };

        using var response = await _client.ExecuteAsync("hash.setnx", key, data, cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("created", out var created) && created.GetBoolean();
    }
}

